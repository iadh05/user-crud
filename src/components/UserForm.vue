<template>
    <v-layout class="flex-row align-center flex-none modal-user-container">
        <v-flex cols="auto">
            <v-icon
                data-cy="update"
                class="mx-1 ma-2"
                v-if="icon && update"
                @click="getUser"
                >mdi-pencil</v-icon
            >
            <v-icon
                class="mx-1 ma-2"
                data-cy="readonly"
                v-else-if="icon && readonly"
                @click="getUser"
                >mdi-file-find</v-icon
            >
            <v-btn
                v-else
                @click="initForm"
                color="primary"
                class="ma-2 white--text text-capitalize mx-1"
            >
                <v-icon left> mdi-plus </v-icon>
                Add User
            </v-btn>

            <v-dialog
                transition="dialog-bottom-transition"
                max-width="600"
                v-model="dialog"
            >
                <v-card>
                    <v-toolbar color="primary" dark>{{
                        displayTitle
                    }}</v-toolbar>
                    <v-card-text class="mt-3">
                        <v-form v-model="validate">
                            <v-text-field
                                label="FirstName"
                                v-model="firstName"
                                :rules="rules.required"
                                :readonly="readonly"
                                data-cy="firstName"
                            ></v-text-field>

                            <v-text-field
                                label="LastName"
                                v-model="lastName"
                                :rules="rules.required"
                                :readonly="readonly"
                                data-cy="lastName"
                            ></v-text-field>

                            <v-text-field
                                v-model="email"
                                label="Email"
                                :rules="rules.email"
                                :readonly="readonly"
                                data-cy="email"
                            ></v-text-field>

                            <v-text-field
                                v-model="address"
                                label="Address"
                                :rules="rules.required"
                                :readonly="readonly"
                                data-cy="address"
                            ></v-text-field>
                        </v-form>
                    </v-card-text>
                    <v-card-actions class="justify-end">
                        <div v-if="!readonly">
                            <v-btn
                                text
                                v-if="update"
                                :disabled="!validate"
                                @click="handleSubmit"
                                data-cy="update-button"
                                >Update</v-btn
                            >
                            <v-btn
                                text
                                v-else
                                :disabled="!validate"
                                @click="handleSubmit"
                                >Save</v-btn
                            >
                        </div>
                        <v-btn text @click="dialog = false">Close</v-btn>
                    </v-card-actions>
                </v-card>
            </v-dialog>
        </v-flex>
    </v-layout>
</template>

<script>
import {mapActions} from "vuex";
import mapFields from "../utils/mapFields";
export default {
    name: "user-form",
    props: ["icon", "userId", "update", "readonly"],
    data() {
        return {
            validate: false,
            dialog: false,
            rules: {
                required: [(v) => !!v || "Field is required"],
                email: [
                    (v) => !!v || "Field is required",
                    (v) => /.+@.+/.test(v) || "Invalid Email",
                ],
            },
        };
    },
    computed: {
        ...mapFields(
            ["firstName", "lastName", "email", "address"],
            "getUser",
            "updateUserFields"
        ),
        displayTitle() {
            if (this.readonly) return "User Profile";
            if (this.update) return "Update User";
            return "Create User";
        },
    },
    methods: {
        ...mapActions([
            "createUser",
            "initFields",
            "getUserById",
            "updateUser",
        ]),

        initForm() {
            this.initFields();
            this.dialog = true;
        },
        getUser() {
            this.getUserById(this.userId);
            this.dialog = true;
        },
        handleSubmit() {
            if (this.update) {
                this.updateUser(this.userId);
            } else {
                this.createUser();
            }
            this.dialog = false;
        },
    },
};
</script>

<style>
.modal-user-container {
    flex: none !important;
}
</style>
