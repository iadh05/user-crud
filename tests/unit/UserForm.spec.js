import Vue from 'vue'
import { mount, createLocalVue, enableAutoDestroy } from '@vue/test-utils'
import UserForm from '@/components/UserForm.vue'
import Vuex from "vuex"
import Vuetify from 'vuetify'


// calls wrapper.destroy() after each test
enableAutoDestroy(afterEach)


const localVue = createLocalVue()

Vue.use(Vuetify);
let actions
let store
let getters;
let state
describe('UserForm.vue', () => {
    beforeEach(() => {
        localVue.use(Vuex)
        actions = {
            updateUserFields: jest.fn(),
            updateUser: jest.fn()
        }
        getters = {
            getUser: (state) => ({ field }) => {
                return state.user[field];
            },

        }
        state = {
            user: {
                firstName: "jhon",
                lastName: "doe",
                email: "test@gmail.com",
                address: '30 street'
            }
        }

        store = new Vuex.Store({
            actions,
            getters,
            state
        })


    })
    it('should show the form element on the user output', () => {
        const wrapper = mount(UserForm, {
            store,
            localVue,
            vuetify: new Vuetify,

            data() {
                return {
                    dialog: true
                }
            },
        })

        expect(wrapper.html()).toMatchSnapshot()
    })
    it("should show update icon", () => {
        const wrapper = mount(UserForm, {
            store,
            localVue,
            vuetify: new Vuetify,

            propsData: {
                update: true,
                icon: true
            }
        })

        expect(wrapper.find("[data-cy='update']").exists()).toBe(true);
    })
    it("should trigger update click", async () => {
        const wrapper = mount(UserForm, {
            store,
            localVue,
            vuetify: new Vuetify,

            propsData: {
                update: true,
                icon: true
            },
            data() {
                return {
                    dialog: true
                }
            }
        })


        const updateClick = wrapper.find("[data-cy='update-button']")
        expect(updateClick.exists()).toBe(true);
        const handleSubmit = jest.fn()
        const firstNameInput = wrapper.find('[data-cy="firstName"]')
        const lastNameInput = wrapper.find('[data-cy="lastName"]')
        const emailInput = wrapper.find('[data-cy="email"]')
        const addressInput = wrapper.find('[data-cy="address"]')
        firstNameInput.setValue('jhon')
        lastNameInput.setValue('doe')
        emailInput.setValue('test@gmail.com')
        addressInput.setValue('30 street')

        await firstNameInput.trigger('[data-cy="firstName"]')
        await lastNameInput.trigger('[data-cy="lastName"]')
        await emailInput.trigger('[data-cy="email"]')
        await addressInput.trigger('[data-cy="address"]')
        await wrapper.setMethods({ handleSubmit })

        expect(updateClick.attributes("disabled")).not.toBe("disabled")

        await updateClick.trigger("click")

        await wrapper.vm.$nextTick()

        expect(handleSubmit).toHaveBeenCalled()
    })
    it("inject props readonly", () => {
        const wrapper = mount(UserForm, {
            store,
            localVue,
            vuetify: new Vuetify,

            propsData: {
                readonly: true,
                icon: true
            }
        })

        expect(wrapper.find("[data-cy='readonly']").exists()).toBe(true);
    })

    it('dispatches "actionInput" when input event value is "input"', async () => {
        const wrapper = mount(UserForm, {
            store,
            localVue,
            vuetify: new Vuetify,

            data() {
                return {
                    dialog: true
                }
            },
        })
        const firstNameInput = wrapper.find('[data-cy="firstName"]')
        const lastNameInput = wrapper.find('[data-cy="lastName"]')
        const emailInput = wrapper.find('[data-cy="email"]')
        const addressInput = wrapper.find('[data-cy="address"]')
        firstNameInput.setValue('jhon')
        lastNameInput.setValue('doe')
        emailInput.setValue('test@gmail.com')
        addressInput.setValue('30 street')

        await firstNameInput.trigger('[data-cy="firstName"]')
        await lastNameInput.trigger('[data-cy="lastName"]')
        await emailInput.trigger('[data-cy="email"]')
        await addressInput.trigger('[data-cy="address"]')
        expect(actions.updateUserFields).toHaveBeenCalled()
    })

})
