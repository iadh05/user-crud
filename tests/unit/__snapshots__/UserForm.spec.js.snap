// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm.vue should show the form element on the user output 1`] = `
<div class="layout flex-row align-center flex-none modal-user-container">
  <div class="flex cols"><button type="button" class="ma-2 white--text text-capitalize mx-1 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-plus theme--light"></i>
            Add User
        </span></button>
    <div role="dialog" class="v-dialog__container">
      <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
        <transition-stub name="dialog-bottom-transition" appear="true">
          <div class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
            <div class="v-card v-sheet theme--light">
              <header class="v-sheet theme--dark v-toolbar primary" style="height: 64px;">
                <div class="v-toolbar__content" style="height: 64px;">Create User</div>
              </header>
              <div class="v-card__text mt-3">
                <form novalidate="novalidate" class="v-form">
                  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-12" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">FirstName</label><input data-cy="firstName" id="input-12" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-15" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">LastName</label><input data-cy="lastName" id="input-15" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-18" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Email</label><input data-cy="email" id="input-18" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-21" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Address</label><input data-cy="address" id="input-21" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
              <div class="v-card__actions justify-end">
                <div><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--text theme--light v-size--default"><span class="v-btn__content">Save</span></button></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default"><span class="v-btn__content">Close</span></button>
              </div>
            </div>
          </div>
        </transition-stub>
      </div>
    </div>
  </div>
</div>
`;
